src/magic_folder/cli.py

```diff
@@ -898,96 +873,6 @@ def list_directory(self, uri):
-        uri = self.node_uri.child(
-            u"uri",
-            filenode_uri.to_string().decode("ascii"),
-        ).to_uri().to_text().encode("ascii")
-
-        with start_action(action_type=u"magic-folder:cli:download", uri=uri):
-            response = yield self.agent.request(
-                b"GET",
-                uri,
-            )
-            if response.code != 200:
-                raise Exception(
-                    "Error response from download endpoint: {code} {phrase}".format(
-                        **vars(response)
-                    ))
-
-        returnValue((yield readBody(response)))
-
-    @inline_callbacks
-    def add_file(self, dirnode_uri, name, uploadable, metadata, overwrite, progress):
-        size = yield uploadable.get_size()
-        contents = b"".join((yield uploadable.read(size)))
-
-        uri = self.node_uri.child(
-            u"uri",
-        ).to_uri().to_text().encode("ascii")
-        action = start_action(
-            action_type=u"magic-folder:cli:add_file:put",
-            uri=uri,
-            size=size,
-        )
-        with action:
-            upload_response = yield self.agent.request(
-                b"PUT",
-                uri,
-                bodyProducer=FileBodyProducer(BytesIO(contents)),
-            )
-
-            if upload_response.code != 200:
-                raise Exception(
-                    "Error response from upload endpoint: {code} {phrase}".format(
-                        **vars(upload_response)
-                    ),
-                )
-
-            filecap = yield readBody(upload_response)
-
-        uri = self.node_uri.child(
-            u"uri",
-            dirnode_uri.to_string().decode("ascii"),
-            u"",
-        ).add(
-            u"t",
-            u"set-children",
-        ).add(
-            u"overwrite",
-            u"true" if overwrite else u"false",
-        ).to_uri().to_text().encode("ascii")
-        action = start_action(
-            action_type=u"magic-folder:cli:add_file:metadata",
-            uri=uri,
-        )
-        with action:
-            response = yield self.agent.request(
-                b"POST",
-                uri,
-                bodyProducer=FileBodyProducer(
-                    BytesIO(
-                        json.dumps({
-                            name: [
-                                u"filenode", {
-                                    "ro_uri": filecap,
-                                    "size": size,
-                                    "metadata": metadata,
-                                },
-                            ],
-                        }).encode("utf-8"),
-                    ),
-                ),
-            )
-            if response.code != 200:
-                raise Exception("Error response from metadata endpoint: {code} {phrase}".format(
-                    **vars(response)
-                ))
-        returnValue(Node(self, from_string(filecap)))
```

Comment:
I think we do need this functionality.  How will files be linked into the DMD?  It seems like they would be linked in with the HTTP API this function wraps.  Is there also another implementation of this functionality somewhere?  I think @meejah was hitting his head against directory interactions in #217 - maybe there's some work there?  Or would this method help with that?

Reply:
This doesn't just "link a file into a DMD" like the `add_file` API (in the "other" client :( ) does .. this accepts an "uploadable" (a Tahoe hold-over interface, I think, since it's not documented?) and uploads it PLUS puts it in the DMD.

Also, I don't think this method as it sits is a good idea/implementation since it sucks the entire file into memory on the first line, so it won't work for anything "big" anyway (and duplicates any functionality for "upload a file").